source/common/upstream/cluster_manager_impl.cc

```diff
@@ -687,17 +687,22 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(const Cluster& cluster, ui
       host_set->healthyHostsPerLocality().clone();
   HostsPerLocalityConstSharedPtr degraded_hosts_per_locality_copy =
       host_set->degradedHostsPerLocality().clone();
+  ExcludedHostVectorConstSharedPtr excluded_hosts_copy(
```

Comment:
Worth doing the @htuch TODO up above here soon in a follow up? This is continues to grow scarier and scarier. :) 

Avoiding the copies here would I think make this function less painful to read and avoid asking for some type of param struct like we have done elsewhere?

Reply:
Wouldn't you still need to do pull out a shared ptr for each of the values and pass them along to the lambda to ensure that we're using consistent values? If we just pass along the host set I think it's possible for the TLS updates to happen concurrently with a host update on the main thread and result in potentially inconsistent values?

I think this whole thing could be simplified pretty easily by reusing the `UpdateHostsParams` param struct that's already being used elsewhere to call `updateHosts`, which would be helped by exposing shared ptrs to the underlying typed arrays. I'll give that a go in this PR and we can decide whether it makes sense to split it out as a different PR

Reply:
> I think this whole thing could be simplified pretty easily by reusing the UpdateHostsParams param struct that's already being used elsewhere to call updateHosts, which would be helped by exposing shared ptrs to the underlying typed arrays. I'll give that a go in this PR and we can decide whether it makes sense to split it out as a different PR

Sorry yeah this is roughly what I meant. I think we can avoid the copies vector and also simplify the copying?